fossas/fossa-cli
# Go: Skip reporting stdlib packages

# Overview

Updates the Go strategy to filter standard library packages from the reported graph.
We use `go list std` to power the filtration; this means that we'll always filter the correct set of standard library dependencies depending on the Go version the user has installed.

`go list std` has been in Go since at least `go1.5`: https://pkg.go.dev/cmd/go@go1.5#hdr-Description_of_package_lists
I can't see documentation older than that but I think it's safe to assume users won't run into issues with this command.

## Acceptance criteria

Go standard library packages are no longer reported for projects.

## Testing plan

I've attached an archive with a reproduction case: [repro.zip](https://github.com/fossas/fossa-cli/files/8578517/repro.zip)

1. Place this project in your `go env GOPATH`.
For me, that's `/Users/kit/go`, so I put it at `/Users/kit/go/src/github.com/kitified/gotest/gomod`.
You may need to rewrite the module name to match your user; sorry, that's Go pre-modules 😬 
2. Set the environment variable `GO111MODULES=off`.
Steps 1 and 2 are required to get the primary tactic, `go list -m -json all`, to fail.
3. Run `fossa analyze -o | jq` on this project. You should see the below (note the `log` and `github.com/kitified/gotest/gomod` deps):
```json
{
  "projects": [
    {
      "type": "gomod",
      "graph": {
        "assocs": [
          [
            1,
            [
              0,
              2
            ]
          ]
        ],
        "deps": [
          {
            "tags": {},
            "name": "github.com/google/uuid",
            "type": "GoType",
            "locations": [],
            "version": {
              "type": "EQUAL",
              "value": "v1.3.0"
            }
          },
          {
            "tags": {},
            "name": "github.com/kitified/gotest/gomod",
            "type": "GoType",
            "locations": [],
            "version": null
          },
          {
            "tags": {},
            "name": "log",
            "type": "GoType",
            "locations": [],
            "version": null
          }
        ],
        "direct": [
          0
        ]
      },
      "path": "/Users/kit/go/src/github.com/kitified/gotest/gomod/"
    }
  ],
  "sourceUnits": [
    {
      "Build": {
        "Dependencies": [
          {
            "imports": [],
            "locator": "go+github.com/google/uuid$v1.3.0"
          },
          {
            "imports": [
              "go+github.com/google/uuid$v1.3.0",
              "go+log$"
            ],
            "locator": "go+github.com/kitified/gotest/gomod$"
          },
          {
            "imports": [],
            "locator": "go+log$"
          }
        ],
        "Imports": [
          "go+github.com/google/uuid$v1.3.0"
        ],
        "Succeeded": true,
        "Artifact": "default"
      },
      "Name": "/Users/kit/go/src/github.com/kitified/gotest/gomod/",
      "Type": "gomod",
      "GraphBreadth": "partial",
      "OriginPaths": [
        "go.mod"
      ],
      "Manifest": "/Users/kit/go/src/github.com/kitified/gotest/gomod/",
      "AdditionalDependencyData": null
    }
  ]
}
```
4. Check out and compile this branch, then run `cabal build && cabal run fossa -- analyze -o | tail -n 1 | jq`.
5. You should see the below (note the `log` and `github.com/kitified/gotest/gomod` deps are no longer showing):
```json
{
  "projects": [
    {
      "type": "gomod",
      "graph": {
        "assocs": [],
        "deps": [
          {
            "tags": {},
            "name": "github.com/google/uuid",
            "type": "GoType",
            "locations": [],
            "version": {
              "type": "EQUAL",
              "value": "v1.3.0"
            }
          }
        ],
        "direct": [
          0
        ]
      },
      "path": "/Users/kit/go/src/github.com/kitified/gotest/gomod/"
    }
  ],
  "sourceUnits": [
    {
      "Build": {
        "Dependencies": [
          {
            "imports": [],
            "locator": "go+github.com/google/uuid$v1.3.0"
          }
        ],
        "Imports": [
          "go+github.com/google/uuid$v1.3.0"
        ],
        "Succeeded": true,
        "Artifact": "default"
      },
      "Name": "/Users/kit/go/src/github.com/kitified/gotest/gomod/",
      "Type": "gomod",
      "GraphBreadth": "complete",
      "OriginPaths": [
        "go.mod"
      ],
      "Manifest": "/Users/kit/go/src/github.com/kitified/gotest/gomod/",
      "AdditionalDependencyData": null
    }
  ]
}
```

## Risks

This makes it to where users will no longer have the Go standard library reported as dependencies in their projects.
For any org that cares about this, this would be a regression. However, as far as we can tell no org on cloud has any issue related to the Go standard library in Core, so we do not believe this would affect anyone.

If it turns out this _does_ affect an org, and they're unwilling to view this in a positive light, we can always add a flag to disable this filtering behavior.

## References

This ticket has been a long time coming but was brought to a head when a user had local deps flagged as missing in [ZD 4098](https://fossa.zendesk.com/agent/tickets/4098), and then I learned that updating the default FOSSA Go version is tricky to do correctly and potentially risky ([Core #7562](https://github.com/fossas/FOSSA/pull/7562)).

Rather than go down that road I thought it best to fix it once and for all.

## Checklist
